packages/eslint-plugin/src/rules/enum-utils/shared.ts

```diff
@@ -33,8 +48,36 @@ export function getEnumTypes(
   typeChecker: ts.TypeChecker,
   type: ts.Type,
 ): ts.Type[] {
-  return tsutils
-    .unionTypeParts(type)
-    .filter(subType => util.isTypeFlagSet(subType, ts.TypeFlags.EnumLiteral))
-    .map(type => getBaseEnumType(typeChecker, type));
+  return getEnumLiterals(type).map(type => getBaseEnumType(typeChecker, type));
+}
+
+/**
+ * Returns the enum key that matches the given literal node, or null if none
+ * match. For example:
+ * ```ts
+ * enum Fruit {
+ *   Apple = 'apple',
+ *   Banana = 'banana',
+ * }
+ *
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'apple') --> 'Fruit.Apple'
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'banana') --> 'Fruit.Banana'
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'orange') --> null
+ * ```
+ */
+export function getEnumKeyForLiteral(
+  enumLiterals: ts.Type[],
+  literal: unknown,
+): string | null {
+  for (const enumLiteral of enumLiterals) {
+    // @ts-expect-error Not sure why `value` is not on `enumLiteral`.
+    if (enumLiteral.value === literal) {
```

Comment:
It's because `ts.Type` is the general type for any type in code. But it seems like this is guaranteed to be the more specific `ts.LiteralType`, since it has a `value` and is from the enum members. (I found that type by going to definition on `ts.Type` and searching for `value: `)

You'll want to put an assertion or type predicate or similar somewhere to indicate that the value being provided to `enumLiterals` is `ts.LiteralType[]`.

Tip: I think this might be helpful...

```plaintext
.filter((subType): subType is ts.LiteralType => ...;
```